utilities/pipelines/resourceRemoval/helper/Remove-GeneralModule.ps1

```diff
@@ -0,0 +1,236 @@
+Optional. The deployment name to use for the removal
+
+.PARAMETER templateFilePath
+Optional. The path to the deployment file
+
+.EXAMPLE
+Remove-GeneralModule -deploymentName 'KeyVault' -resourceGroupName 'validation-rg'
+
+Remove a virtual WAN with deployment name 'keyvault-12345' from resource group 'validation-rg'
+#>
+function Remove-GeneralModule {
+
+    [CmdletBinding(SupportsShouldProcess)]
+    param (
+        [Parameter(Mandatory = $false)]
+        [string] $resourceGroupName,
+
+        [Parameter(Mandatory = $true)]
+        [string] $deploymentName,
+
+        [Parameter(Mandatory = $true)]
+        [string] $templateFilePath,
+
+        [Parameter(Mandatory = $false)]
+        [int] $searchRetryLimit = 40,
+
+        [Parameter(Mandatory = $false)]
+        [int] $searchRetryInterval = 60
+    )
+
+    begin {
+        Write-Debug ('{0} entered' -f $MyInvocation.MyCommand)
+
+        # Load helper
+        . (Join-Path $PSScriptRoot 'Remove-Resource.ps1')
+    }
+
+    process {
+
+        #####################
+        ## Process Removal ##
+        #####################
+        Write-Verbose ('Handling resource removal with deployment name [{0}]' -f $deploymentName) -Verbose
+
+        # Gather deployments
+        # ==================
+        if ((Split-Path $templateFilePath -Extension) -eq '.bicep') {
+            # Bicep
+            $bicepContent = Get-Content $templateFilePath
+            $bicepScope = $bicepContent | Where-Object { $_ -like '*targetscope =*' }
+            if (-not $bicepScope) {
+                $deploymentScope = 'resourceGroup'
+            } else {
+                $deploymentScope = $bicepScope.ToLower().Replace('targetscope = ', '').Replace("'", '').Trim()
+            }
+        } else {
+            # ARM
+            $armSchema = (ConvertFrom-Json (Get-Content -Raw -Path $templateFilePath)).'$schema'
+            switch -regex ($armSchema) {
+                '\/deploymentTemplate.json#$' { $deploymentScope = 'resourceGroup' }
+                '\/subscriptionDeploymentTemplate.json#$' { $deploymentScope = 'subscription' }
+                '\/managementGroupDeploymentTemplate.json#$' { $deploymentScope = 'managementGroup' }
+                '\/tenantDeploymentTemplate.json#$' { $deploymentScope = 'tenant' }
+                Default { throw "[$armSchema] is a non-supported ARM template schema" }
+            }
+        }
+
+        $searchRetryCount = 1
+        while (-not ($deployments = Get-DeploymentByName -name $deploymentName -scope $deploymentScope -resourceGroupName $resourceGroupName -ErrorAction 'SilentlyContinue') -and $searchRetryCount -le $searchRetryLimit) {
+            Write-Verbose ('Did not to find deployments by name [{0}] in scope [{1}]. Retrying in [{2}] seconds [{3}/{4}]' -f $deploymentName, $deploymentScope, $searchRetryInterval, $searchRetryCount, $searchRetryLimit) -Verbose
+            Start-Sleep $searchRetryInterval
+            $searchRetryCount++
+        }
```

Comment:
Otherwise I have to duplicate the code. I.e. I'd need to run it before the loop, and then again inside the loop if the condition does not hold true. Any suggestions?

Reply:
Not 100% sure, but would it not be this?

```
do {
    $deployments = Get-DeploymentByName -name $deploymentName -scope $deploymentScope -resourceGroupName $resourceGroupName -ErrorAction 'SilentlyContinue'
    if (-not $deployments) {
        break
    }
    Write-Verbose ('Did not to find deployments by name [{0}] in scope [{1}]. Retrying in [{2}] seconds [{3}/{4}]' -f $deploymentName, $deploymentScope, $searchRetryInterval, $searchRetryCount, $searchRetryLimit) -Verbose
    Start-Sleep $searchRetryInterval
    $searchRetryCount++
} while ($searchRetryCount -le $searchRetryLimit)
```

Reply:
You are a smart man @MariusStorhaug

Reply:
Updated